src/V3Options.cpp

```diff
@@ -104,6 +106,223 @@ class V3OptionsImp final {
     ~V3OptionsImp() = default;
 };
 
+//######################################################################
+// V3 OptionParser
+
+class V3OptionsParser final {
+public:
+    // TYPES
+    using CallbackFn = std::function<void(const char*, const char*)>;
+    enum enType : uint8_t {  // Type of target variable
+        TYPE_BOOL,  // bool
+        TYPE_BOOLOPT,  // VOptionBool
+        TYPE_INT,  // int
+        TYPE_STRING,  // string
+        TYPE_CB  // CallbackFn
+    };
+    enum enAction : uint8_t {
+        ACT_SET,  // '-opt' for bool-ish, '-opt val' for int and string
+        ACT_ONOFF,  // '-opt' or '-no-opt' for bool-ish type
+        ACT_CB_CALL,  // '-opt' for callback
+        ACT_CB_ONOFF,  // '-opt' or '-no-opt' for callback
+        ACT_CB_VAL,  // '-opt val' for callback
+        ACT_CB_PARTIAL_MATCH,  // '-opt-abc' for callback
+        ACT_CB_PARTIAL_MATCH_VAL  // '-opt-abc val' for callback
+    };
+    template <enAction ACT> struct ActionBase VL_NOT_FINAL {
+        static constexpr enAction m_act = ACT;
+    };
+    // Function selectors for AppendHelper
+    using ActSet = ActionBase<ACT_SET>;
+    using ActOnOff = ActionBase<ACT_ONOFF>;
+    using ActCbCall = ActionBase<ACT_CB_CALL>;
+    using ActCbOnOff = ActionBase<ACT_CB_ONOFF>;
+    using ActCbVal = ActionBase<ACT_CB_VAL>;
+    using ActCbPartialMatch = ActionBase<ACT_CB_PARTIAL_MATCH>;
+    using ActCbPartialMatchVal = ActionBase<ACT_CB_PARTIAL_MATCH_VAL>;
+
+    // Entry for each option
+    struct Entry final {
+        enType m_type;
+        enAction m_act;  // Action to do
+        union Val {
+            bool* m_boolp;
+            VOptionBool* m_boolOptp;
+            int* m_intp;
+            string* m_strp;
```

Comment:
Alternatively just a void* and deal with interpretation on the way out?  You already need to have a case statement.